File: qingyu_command_example_ws/src/qingyu_command_publisher_cpp/src/qingyu_command_pub.cpp
#include <chrono>
#include <string>

#include <rclcpp/rclcpp.hpp>

#include "qingyu_api/msg/qingyu_command.hpp"

using namespace std::chrono_literals;

class QingyuCommandPublisher final : public rclcpp::Node
{
public:
  explicit QingyuCommandPublisher(const rclcpp::NodeOptions & options)
  : rclcpp::Node("qingyu_api", "/qingyu_api", options)
  {
    topic_name_ = this->declare_parameter<std::string>("topic", "/qingyu_api");
    publish_hz_ = this->declare_parameter<double>("publish_hz", 50.0);

    this->declare_parameter<double>("x", 0.0);
    this->declare_parameter<double>("y", 0.0);
    this->declare_parameter<double>("z", 0.0);
    this->declare_parameter<double>("roll", 0.0);
    this->declare_parameter<double>("pitch", 0.0);
    this->declare_parameter<double>("yaw", 0.0);
    this->declare_parameter<int>("cmd_mode", 1);
    this->declare_parameter<double>("a1", 0.0);
    this->declare_parameter<double>("b1", 0.0);
    this->declare_parameter<double>("a2", 0.0);
    this->declare_parameter<double>("b2", 0.0);
    this->declare_parameter<double>("a3", 0.0);
    this->declare_parameter<double>("b3", 0.0);
    this->declare_parameter<bool>("enable", true);

    publisher_ = this->create_publisher<qingyu_api::msg::QingyuCommand>(topic_name_, rclcpp::QoS(10));

    const auto period = publish_hz_ > 0.0 ? std::chrono::duration<double>(1.0 / publish_hz_) : 20ms;
    timer_ = this->create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(period),
      std::bind(&QingyuCommandPublisher::onTimer, this));
  }

private:
  void onTimer()
  {
    qingyu_api::msg::QingyuCommand msg;
    msg.x = static_cast<float>(this->get_parameter("x").as_double());
    msg.y = static_cast<float>(this->get_parameter("y").as_double());
    msg.z = static_cast<float>(this->get_parameter("z").as_double());
    msg.roll = static_cast<float>(this->get_parameter("roll").as_double());
    msg.pitch = static_cast<float>(this->get_parameter("pitch").as_double());
    msg.yaw = static_cast<float>(this->get_parameter("yaw").as_double());

    msg.cmd_mode = static_cast<uint8_t>(this->get_parameter("cmd_mode").as_int());

    msg.a1 = static_cast<float>(this->get_parameter("a1").as_double());
    msg.b1 = static_cast<float>(this->get_parameter("b1").as_double());
    msg.a2 = static_cast<float>(this->get_parameter("a2").as_double());
    msg.b2 = static_cast<float>(this->get_parameter("b2").as_double());
    msg.a3 = static_cast<float>(this->get_parameter("a3").as_double());
    msg.b3 = static_cast<float>(this->get_parameter("b3").as_double());
    msg.enable = this->get_parameter("enable").as_bool();

    publisher_->publish(msg);
  }

  std::string topic_name_;
  double publish_hz_{50.0};
  rclcpp::Publisher<qingyu_api::msg::QingyuCommand>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<QingyuCommandPublisher>(rclcpp::NodeOptions{}));
  rclcpp::shutdown();
  return 0;
}

File: qingyu_command_example_ws/src/qingyu_command_publisher_py/qingyu_command_publisher_py/qingyu_command_pub.py
from __future__ import annotations

import rclpy
from rclpy.node import Node

from qingyu_api.msg import QingyuCommand


class QingyuCommandPublisher(Node):
    def __init__(self) -> None:
        super().__init__("qingyu_api", namespace="/qingyu_api")

        self.declare_parameter("topic", "/qingyu_api")
        self.declare_parameter("publish_hz", 50.0)

        self.declare_parameter("x", 0.0)
        self.declare_parameter("y", 0.0)
        self.declare_parameter("z", 0.0)
        self.declare_parameter("roll", 0.0)
        self.declare_parameter("pitch", 0.0)
        self.declare_parameter("yaw", 0.0)
        self.declare_parameter("cmd_mode", 1)
        self.declare_parameter("a1", 0.0)
        self.declare_parameter("b1", 0.0)
        self.declare_parameter("a2", 0.0)
        self.declare_parameter("b2", 0.0)
        self.declare_parameter("a3", 0.0)
        self.declare_parameter("b3", 0.0)
        self.declare_parameter("enable", True)

        self._topic_name = self.get_parameter("topic").get_parameter_value().string_value
        publish_hz = self.get_parameter("publish_hz").get_parameter_value().double_value
        period_sec = 1.0 / publish_hz if publish_hz > 0.0 else 0.02

        self._publisher = self.create_publisher(QingyuCommand, self._topic_name, 10)
        self._timer = self.create_timer(period_sec, self._on_timer)

    def _on_timer(self) -> None:
        msg = QingyuCommand()
        msg.x = float(self.get_parameter("x").value)
        msg.y = float(self.get_parameter("y").value)
        msg.z = float(self.get_parameter("z").value)
        msg.roll = float(self.get_parameter("roll").value)
        msg.pitch = float(self.get_parameter("pitch").value)
        msg.yaw = float(self.get_parameter("yaw").value)

        cmd_mode = int(self.get_parameter("cmd_mode").value)
        if cmd_mode < 0:
            cmd_mode = 0
        elif cmd_mode > 255:
            cmd_mode = 255
        msg.cmd_mode = cmd_mode

        msg.a1 = float(self.get_parameter("a1").value)
        msg.b1 = float(self.get_parameter("b1").value)
        msg.a2 = float(self.get_parameter("a2").value)
        msg.b2 = float(self.get_parameter("b2").value)
        msg.a3 = float(self.get_parameter("a3").value)
        msg.b3 = float(self.get_parameter("b3").value)
        msg.enable = bool(self.get_parameter("enable").value)

        self._publisher.publish(msg)


def main() -> None:
    rclpy.init()
    node = QingyuCommandPublisher()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()

